File: webapps/frontend/app/viewmodels/services/vdisk.js
// Copyright (C) 2017 iNuron NV
//
// This file is part of Open vStorage Open Source Edition (OSE),
// as available from
//
//      http://www.openvstorage.org and
//      http://www.openvstorage.com.
//
// This file is free software; you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
// as published by the Free Software Foundation, in version 3 as it comes
// in the LICENSE.txt file of the Open vStorage OSE distribution.
//
// Open vStorage is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY of any kind.
/*global define */
/**
 * Service to help with vdisk related tasks
 */
define([
    'jquery', 'knockout',
    'ovs/api', 'ovs/shared'
], function ($, ko,
             api, shared) {
    function VDiskService() {
        var self = this;

        /**
         * Loads the vdisk
         * @param guid: guid of the vdisk
         * @returns {Promise<T>}
         */
        self.loadVDisk = function (guid) {
            return api.get('vdisks/' + guid)
        };


        /**
         * Get the vdisk handle
         * @param options: options of the vdisk
         * @returns {Promise<T>}
         */
        self.loadVDisksHandle = function (options) {
            return api.get('vdisks', { queryparams: options })
                .then(shared.tasks.wait)
        };


        /**
         * Loads the config params of the vdisk
         * @param guid: Additional query params.
         * @returns {Promise<T>}
         */
        self.loadConfig = function (guid) {
            return api.get('vdisks/' + guid + '/get_config_params')
                .then(shared.tasks.wait)
        };
        /**
         * Update the configuration parameters of the VDisk
         * @param guid: Guid of the VDisk
         * @param new_config_params: New config params
         * @returns {Promise<T>}
         */
        self.setConfig = function (guid, new_config_params) {
                return api.post('vdisks/' + guid + '/set_config_params', {
                    data: { new_config_params: new_config_params}
                })
                .then(shared.tasks.wait)
        };
        /**
         * Set the VDisk as template
         * @param guid: Guid of the VDisk
         * @returns {Promise<T>}
         */
        self.setAsTemplate = function (guid){
            return api.post('vdisks/' + guid + '/set_as_template')
                .then(shared.tasks.wait)
        };
        /**
         * Scrub the VDisk
         * @param guid: Guid of the VDisk
         * @returns {Promise<T>}
         */
        self.scrub = function (guid){
            return api.post('vdisks/' + guid + '/scrub')
                .then(shared.tasks.wait)
        };
        /**
         * Remove a snapshot from the VDisk
         * @param guid: Guid of the VDisk
         * @param snapshotid: ID of the snapshot
         * @returns {Promise<T>}
         */
        self.removeSnapshot = function (guid, snapshotid){
            return api.post('vdisks/' + guid + '/remove_snapshot', {
                    data: { snapshot_id: snapshotid }
                }).then(shared.tasks.wait)
        };
        /**
         * Remove a VDisk
         * @param guid: Guid of the VDisk
         * @returns {Promise<T>}
         */
        self.removeVDisk = function (guid){
            return api.del('vdisks/' + guid)
                .then(shared.tasks.wait)
        };
        /**
         * Restart a VDisk
         * @param guid: Guid of the VDisk
         * @returns {Promise<T>}
         */
        self.restart = function (guid){
            return api.post('vdisks/' + guid + '/restart')
                .then(shared.tasks.wait)
        };
        /**
         * Cleans the device name of a VDisk
         * @param name: Name of the VDisk
         * @returns {string}
         */
        self.cleanDeviceName = function(name) {
            var cleaned = name.replace(/^(\/)+|(\/)+$/g, '').replace(/ /g,"_").replace(/[^a-zA-Z0-9-_\.\/]+/g, "");
            while (cleaned.indexOf('//') > -1) {
                cleaned = cleaned.replace(/\/\//g, '/');
            }
            if (cleaned.length > 4 && cleaned.slice(-4) === '.raw') {
                return cleaned;
            }
            return cleaned + '.raw';
        }
    }
    return new VDiskService();

});

File: ovs/constants/vdisk.py
# Copyright (C) 2018 iNuron NV
#
# This file is part of Open vStorage Open Source Edition (OSE),
# as available from
#
#      http://www.openvstorage.org and
#      http://www.openvstorage.com.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
# as published by the Free Software Foundation, in version 3 as it comes
# in the LICENSE.txt file of the Open vStorage OSE distribution.
#
# Open vStorage is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY of any kind.

"""
VDisk Constants module. Contains constants related to vdisks
"""
import os

# General
LOCK_NAMESPACE = 'ovs_locks'

# Scrub related
SCRUB_VDISK_LOCK = '{0}_{{0}}'.format(LOCK_NAMESPACE)  # Second format is the vdisk guid
SCRUB_VDISK_EXCEPTION_MESSAGE = 'VDisk is being scrubbed. Unable to remove snapshots at this time'

# Snapshot related
# Note: the scheduled task will always skip the first 24 hours before enforcing the policy
SNAPSHOT_POLICY_DEFAULT = [# one per day for rest of the week and opt for a consistent snapshot for the first day
                           {'nr_of_snapshots': 7, 'nr_of_days': 7, 'consistency_first': True, 'consistency_first_on': [1]},
                           # One per week for the rest of the month
                           {'nr_of_snapshots': 3, 'nr_of_days': 21}]
SNAPSHOT_POLICY_LOCATION = os.path.join(os.path.sep, 'ovs', 'cluster', 'snapshot_retention_policy')
